### ISSUE TITLE
Online Help System

### ISSUE THREAD

**Issue by ethercrow** [yi-editor/yi#94]
_Original author: aleks.di...@gmail.com (March 16, 2009 15:38:23)_

This is an issue which has been discussed already on the mailing list. As
per Jean-Philippe's request, I'm putting it on the tracker, so we can make
it easier to follow progress, and/or opinions.

This is similar to #249, albeit a bit more general. I'm proposing the
following enhancement:

We need an online help system, similar to Emacs' C-h and Vim's :help. I do
not think that it would be a good idea to use haddock strings and/or let
functions self-document directly in the code.

(below is some text copied from an earlier email to the list)

Vim (and Emacs, I think) come with a special editor mode for
documentation. That means, there are some plain text files, which
follow a certain markup standard, and are indexed by the editor's
online help system. When you search for something, the editor opens up
the relevant file in a new buffer, at the relevant line, and maybe
highlights what you've been looking for. In Vim at least, (I wouldn't
know about Emacs) it also works this way for simple things like `gq` or
`G`; i.e. everything is documented that way.

This decoupling of functionality/documentation certainly has its drawbacks,
but I think it's overall a good idea. I think people should be able to work
on the docs without even looking at the code.

Of course this method will place us in danger of the typical docs/code
race, but that's always the case when you add documentation: keeping
the docs up-to-date becomes a major issue.

Some ideas for a realization:
- use markdown for the documentation since it's widely known and rather easy
  to parse. We'd have to use a subset of md, since a terminal can't show all
  the possible markups. I'm mainly thinking about code samples and links.
  Using the pandoc package, we could even optionally display the help as
  html pages in the browser.
- create a syntax mode for those help files, that allows the user to jump on
  links, similar to Vim's help system or info pages.
- create an indexing system (is there some haskell package that already does
  that?) and make it searchable via some function.

I think that for the two most popular key maps (Emacs and Vim), we could
just reuse the native docs of the respective editor, since Yi's behaviour is
largely similar.

_Original issue: http://code.google.com/p/yi-editor/issues/detail?id=253_

**Comment by Araeos** [comment]
I agree that self documentation would improve the experience for beginners and experienced users alike.
Things that should be integrated in Yi in my opinion are:
- Tutorial document. Can be displayed on first startup when no configuration is present.
- Modules can describe their purpose and usage (user-provided and pre-packaged alike).
- Functions, data types and similiar objects should readily provide their documentation within Yi to make hacking/configuration quick.

Given the points above, there are 2 types of documentation.
On one hand the tutorial and module documentation is not directly related to the code implementation and can provide better overview and understanding.
On the other, how to use specific functions and types is directly related to the implementation and most programmers provide such documentation during development (haddock) already.i am certain the minority of people would in addtion maintain a separate markdown file for the same purpose.

In conclusion *interlinked* documentation systems are needed. One is separate from the code and might use a certain markdown format (this would effectively be an info-like system).
The other is generated from the haddock and could be updated when the code is compiled so that user code is fully documented and it is easy to maintain a good documentation.
